### PR TITLE
Add local registry to existing test framework

### PR DESCRIPTION
**Pull request by kimikowang**

### CODE REVIEW

File: tests/e2e/framework/kubernetes.go [hunk 1]
```diff
@@ -58,6 +58,11 @@ const (
 	defaultGalleyConfigValidatorFile = "istio-galley-config-validator.yaml"
 	maxDeploymentRolloutTime         = 480 * time.Second
 	mtlsExcludedServicesPattern      = "mtlsExcludedServices:\\s*\\[(.*)\\]"
+	// Default values for local test env setup
+	localRegistryFile      = "tests/util/localregistry/localregistry.yaml"
+	localRegistryHub       = "10.10.0.2:5000"
```
Comment (costinm) [inline tests/e2e/framework/kubernetes.go]: Env variable ? Which IP is this ?

File: tests/e2e/README.md [hunk 2]
```diff
@@ -96,6 +96,20 @@ make push
 
 ```
 
+**Option 1.5:** Build your own images and use an in-cluster registry
+
+```
+# Set up env vars in .istiorc.mk
+cat .istiorc.mk
+LOCALREG=true
```
Comment (hklai) [inline tests/e2e/README.md]: Can use just do this automatically using make arguments?

something like "make run_registry_in_cluster=true test e2e_simple"?

File: tests/e2e/README.md [hunk 3]
```diff
@@ -96,6 +96,20 @@ make push
...
+HUB="10.10.0.2:5000"
+TAG=latest
+```
+# Setting the LOCALREG var will notify the e2e test targets to bring up the local registry within cluster and
```
Comment (hklai) [inline tests/e2e/README.md]: Perhaps a bit more info about why this is a good idea.

File: tests/e2e/README.md [hunk 4]
```diff
@@ -96,6 +96,20 @@ make push
...
+# build/push images onto it.
+
+# After the test is done, the registry will be cleanup by default.
```
Comment (hklai) [inline tests/e2e/README.md]: How not to clean up the images and registry?

File: tests/istio.mk [hunk 5]
```diff
@@ -69,20 +72,27 @@ DEFAULT_UPGRADE_E2E_ARGS += --base_version=${LAST_RELEASE}
 DEFAULT_UPGRADE_E2E_ARGS += --target_version=""
 UPGRADE_E2E_ARGS ?= ${DEFAULT_UPGRADE_E2E_ARGS}
 
+# deploy local registry pod, build and push images
+localregistry_setup:
+	ifeq (${LOCALREG},true)
+	kubectl apply -f ${LOCALREG_FILE}
+	GOOS=linux docker docker.push
```
Comment (hklai) [inline tests/istio.mk]: Any need to probe the readiness of registry before pushing?

File: tests/e2e/framework/kubernetes.go [hunk 6]
```diff
@@ -58,24 +58,30 @@ const (
...
+	caTag            = flag.String("ca_tag", os.Getenv("TAG"), "Ca tag")
+	galleyHub        = flag.String("galley_hub", os.Getenv("HUB"), "Galley hub")
+	galleyTag        = flag.String("galley_tag", os.Getenv("TAG"), "Galley tag")
+	localRegistryHub = flag.String("local_registry_hub", os.Getenv("HUB"), "Local Registry hub")
+	localRegistryTag = flag.String("local_registry_tag", os.Getenv("TAG"), "Local Registry tag")
+	authEnable       = flag.Bool("auth_enable", false, "Enable auth")
+	rbacEnable       = flag.Bool("rbac_enable", true, "Enable rbac")
+	localCluster     = flag.Bool("use_local_cluster", false,
 		"Whether the cluster is local or not (i.e. the test is running within the cluster). If running on minikube, this should be set to true.")
+	localRegistry             = flag.Bool("use_local_registry", os.Getenv("LOCALREG"), "Use in-cluster docker registry for testing")
```
Comment (nmittler) [inline tests/e2e/framework/kubernetes.go]: I think the default should be true.  In other words, by default we'll push a local registry (and local images) to the k8s cluster. Then the question becomes: which images do we use? I suggest that if Hub/Tag are unspecified for a given image (i.e. pilot, mixer, etc.), then we default to using the image that we pushed to the local registry.

WDYT?

File: tests/e2e/README.md (hunk 4 above)
Comment (kimikowang) [reply tests/e2e/README.md]: If flag `skip-cleanup` is specified.

**Comment by hklai** [comment]
@rshriram This is part of the swat effort. Let's not delete it for now.

@nmittler @ozevren @kimikowang @gargnupur @JimmyCYJ Now we have multiple branches on different things. Shall we create a new swat branch and move all works there? Since different people are working on different things, we should not step on others anyway.

**Comment by kimikowang** [comment]
The new PR I am going to submit will make use of local test setup in another branch. I think merging them together would be a good idea.

**Comment by hklai** [comment]
@ozevren Since most of the changes are not related to the product itself, we can generally merge to master (maybe periodically) as long as they are in reasonable state, and is not enabled until the part is ready. They don't necessarily all land at the same time.

**Pull request closed without merging**